### PR TITLE
cunumeric: implement `np.diff`

### PR DESCRIPTION
**Pull request by rohany**
This commit implements some necessary cases of `np.diff`.

Signed-off-by: Rohan Yadav <rohany@alumni.cmu.edu>

### CODE REVIEW

File: cunumeric/module.py [hunk 1]
```diff
@@ -5135,6 +5135,120 @@ def nancumsum(
...
+    array([[-1,  2,  0, -2]])
+    >>> x = np.arange('1066-10-13', '1066-10-16', dtype=np.datetime64)
+    >>> np.diff(x)
+    array([1, 1], dtype='timedelta64[D]')
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
+    """
+    if n == 0:
+        return a
```
Comment (bryevdv) [inline cunumeric/module.py:5219]: Do we want to return the exact same object? Numpy returns a copy
```python
In [7]: a = np.array([])

In [8]: np.diff(a) is a
Out[8]: False
```

File: tests/integration/test_diff.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+@pytest.mark.parametrize("n", [1, 2, 3])
+def test_diff(n):
+    num = np.random.random(100)
+    cun = cn.array(num)
+
+    res_num = np.diff(num, n=n)
+    res_cn = cn.diff(cun, n=n)
+
+    assert allclose(res_num, res_cn)
+
```
Comment (bryevdv) [inline tests/integration/test_diff.py:32]: empty edge case should get a test

Comment (bryevdv) [reply tests/integration/test_diff.py:32]: Also the case where `n` is greater than the length of the input

File: cunumeric/module.py [hunk 3]
```diff
@@ -5135,6 +5135,120 @@ def nancumsum(
...
+    if nd == 0:
+        raise ValueError(
+            "diff requires input that is at least one dimensional"
+        )
+    axis = normalize_axis_index(axis, nd)
+
+    if prepend is not None or append is not None:
+        # We don't have prepend and append yet because they require
+        # broadcast_to, which is in progress:
+        # https://github.com/nv-legate/cunumeric/pull/458.
```
Comment (bryevdv) [inline cunumeric/module.py:5233]: FYI broadcast was merged: https://github.com/nv-legate/cunumeric/pull/759

File: cunumeric/module.py [hunk 4]
```diff
@@ -5135,6 +5135,120 @@ def nancumsum(
...
+    array([[2, 3, 4],
+           [5, 1, 2]])
+    >>> np.diff(x, axis=0)
+    array([[-1,  2,  0, -2]])
+    >>> x = np.arange('1066-10-13', '1066-10-16', dtype=np.datetime64)
+    >>> np.diff(x)
+    array([1, 1], dtype='timedelta64[D]')
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
```
Comment (bryevdv) [inline cunumeric/module.py:5216]: Typically docstring section headers have a line of space in between (here and headers above)
```suggestion
    array([1, 1], dtype='timedelta64[D]')
    
    Availability
    --------
    Multiple GPUs, Multiple CPUs
```

File: cunumeric/module.py [hunk 5]
```diff
@@ -5135,6 +5135,120 @@ def nancumsum(
...
+    axis : int, optional
+        The axis along which the difference is taken, default is the
+        last axis.
+    prepend, append : array_like, optional
+        Values to prepend or append to `a` along axis prior to
+        performing the difference.  Scalar values are expanded to
+        arrays with length 1 in the direction of axis and the shape
+        of the input array in along all other axes.  Otherwise the
+        dimension and shape must match `a` except along axis.
+        .. versionadded:: 1.16.0
```
Comment (bryevdv) [inline cunumeric/module.py:5167]: Numpy `versionadded` not really relevant for us I don't think

File: tests/integration/test_diff.py (hunk 2 above)
Comment (bryevdv) [reply tests/integration/test_diff.py:32]: And also need multi-dimensional / `axis` arg coverage

**Review by bryevdv — CHANGES_REQUESTED** [review]
Looking good just a few comments

File: tests/integration/test_diff.py (hunk 2 above)
Comment (rohany) [reply tests/integration/test_diff.py:55]: updated

File: cunumeric/module.py (hunk 3 above)
Comment (rohany) [reply cunumeric/module.py]: thanks, added the prepend and append cases

**Comment by rohany** [comment]
Updated, @bryevdv PTAL

**Comment by bryevdv** [comment]
One linter error @rohany 
```
cunumeric/module.py:6374:10: F821 undefined name 'not_equal'
```

**Comment by rohany** [comment]
woops, fixed the missing import

**Comment by bryevdv** [comment]
\ok to test

File: tests/integration/test_diff.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import numpy as np
+import pytest
+from utils.comparisons import allclose
+
+import cunumeric as cn
```
Comment (manopapad) [inline tests/integration/test_diff.py:20]: I think we've tried to stick to `import cunumeric as num` in the tests. I am personally more of a fan of `cn`, but since `num` was more common we just switched to `num` everywhere.

File: cunumeric/module.py [hunk 7]
```diff
@@ -6242,6 +6242,142 @@ def nansum(
...
 
 
+@add_boilerplate("a")
+def diff(
+    a: ndarray,
+    n: int = 1,
+    axis: int = -1,
+    prepend: Any = None,
+    append: Any = None,
+) -> ndarray:
```
Comment (manopapad) [inline cunumeric/module.py:6252]: With this you can skip the `asanyarray` below, you'll already have an `ndarray` if it was not `None`.

```suggestion
@add_boilerplate("a", "prepend", "append")
def diff(
    a: ndarray,
    n: int = 1,
    axis: int = -1,
    prepend: ndarray | None = None,
    append: ndarray | None = None,
) -> ndarray:
```

File: cunumeric/module.py [hunk 8]
```diff
@@ -6242,6 +6242,142 @@ def nansum(
...
+    array([ 1,  2,  3, -7])
+    >>> np.diff(x, n=2)
+    array([  1,   1, -10])
+    >>> x = np.array([[1, 3, 6, 10], [0, 5, 6, 8]])
+    >>> np.diff(x)
+    array([[2, 3, 4],
+           [5, 1, 2]])
+    >>> np.diff(x, axis=0)
+    array([[-1,  2,  0, -2]])
+    >>> x = np.arange('1066-10-13', '1066-10-16', dtype=np.datetime64)
```
Comment (manopapad) [inline cunumeric/module.py:6317]: cuNumeric doesn't support datetime

File: cunumeric/module.py [hunk 9]
```diff
@@ -6242,6 +6242,142 @@ def nansum(
...
+        dimension and shape must match `a` except along axis.
+    Returns
+    -------
+    diff : ndarray
+        The n-th differences. The shape of the output is the same as `a`
+        except along `axis` where the dimension is smaller by `n`. The
+        type of the output is the same as the type of the difference
+        between any two elements of `a`. This is the same as the type of
+        `a` in most cases. A notable exception is `datetime64`, which
+        results in a `timedelta64` output array.
```
Comment (manopapad) [inline cunumeric/module.py:6282]: Datetime not supported by cuNumeric
```suggestion
        `a` in most cases.
```

File: cunumeric/module.py [hunk 10]
```diff
@@ -6242,6 +6242,142 @@ def nansum(
...
+    See Also
+    --------
+    gradient, ediff1d, cumsum
```
Comment (manopapad) [inline cunumeric/module.py:6285]: Typically want to point to the original NumPy docs.
```suggestion
    numpy.diff
```

File: cunumeric/module.py [hunk 11]
```diff
@@ -6242,6 +6242,142 @@ def nansum(
...
+        )
+    axis = normalize_axis_index(axis, nd)
+
+    combined = []
+    if prepend is not None:
+        prepend = np.asanyarray(prepend)
+        if prepend.ndim == 0:
+            shape = list(a.shape)
+            shape[axis] = 1
+            prepend = np.broadcast_to(prepend, tuple(shape))
```
Comment (manopapad) [inline cunumeric/module.py:6343]: We have this now in cuNumeric, no need to defer to NumPy.
```suggestion
            prepend = broadcast_to(prepend, tuple(shape))
```

File: cunumeric/module.py [hunk 12]
```diff
@@ -6242,6 +6242,142 @@ def nansum(
...
+        combined.append(prepend)
+
+    combined.append(a)
+
+    if append is not None:
+        append = np.asanyarray(append)
+        if append.ndim == 0:
+            shape = list(a.shape)
+            shape[axis] = 1
+            append = np.broadcast_to(append, tuple(shape))
```
Comment (manopapad) [inline cunumeric/module.py:6353]:
```suggestion
            append = broadcast_to(append, tuple(shape))
```

File: cunumeric/module.py [hunk 13]
```diff
@@ -6242,6 +6242,142 @@ def nansum(
...
+        combined.append(append)
+
+    if len(combined) > 1:
+        a = np.concatenate(combined, axis)
```
Comment (manopapad) [inline cunumeric/module.py:6357]:
```suggestion
        a = concatenate(combined, axis)
```

File: cunumeric/module.py [hunk 14]
```diff
@@ -6242,6 +6242,142 @@ def nansum(
...
+
+    # Diffing with n > shape results in an empty array. We have
+    # to handle this case explicitly as our slicing routines raise
+    # an exception with out-of-bounds slices, while NumPy's dont.
+    if a.shape[axis] <= n:
+        shape = list(a.shape)
+        shape[axis] = 0
+        return np.empty(shape=shape, dtype=a.dtype)
```
Comment (manopapad) [inline cunumeric/module.py:6365]:
```suggestion
        return empty(shape=shape, dtype=a.dtype)
```

**Review by manopapad — COMMENTED** [review]
Have some nits from me too

**Comment by rohany** [comment]
Updated, ptal.

**Review by manopapad — APPROVED** [review]

**Comment by rohany** [comment]
@bryedvdv can you accept / merge it?

**Review by bryevdv — APPROVED** [review]
